File: python/tests/test_ansi.py
"""Parity tests for kimix_native.stream ANSI strip (plan 003).

Compares the native stripper against the pure-Python reference
``_compat_filter_output`` (mirror of tools/common.py) on:
- hand-built vectors: OSC hyperlinks, SGR, cursor moves, DCS/PM/APC, Fe,
  partial-escape-at-eof, CRLF/CR mixes
- a 200-byte sample split at EVERY byte offset (feed at split, feed rest,
  flush) == one-shot, via the native LineProcessor
- byte-by-byte feeding == one-shot
- the KIMIX_NATIVE_STREAM=0 fallback toggle
"""

import random

import pytest

import kimix_native
from kimix_native import stream

ESC = "\x1b"

# Corpus of ANSI-heavy samples (mirrors the kimi-agent `regex` behavior).
ANSI_CORPUS = [
    f"{ESC}[31mred{ESC}[0m",
    f"{ESC}]0;title\x07hello",
    f"{ESC}]0;title{ESC}\\hello",
    f"{ESC}Pabc\x07x",
    f"{ESC}Pabc{ESC}[31mx",
    f"{ESC}_ab\x07",
    f"{ESC}_",
    f"{ESC}P",
    f"{ESC}^",
    f"{ESC}]abc",
    f"{ESC}]a{ESC}[0m",
    f"{ESC}]a{ESC}{ESC}\\",
    f"{ESC}[1 2m",
    f"{ESC}[31m",
    f"{ESC}[?25l",
    f"{ESC}[2J{ESC}[H",
    f"{ESC}(0",
    f"{ESC}M",
    f"{ESC}-",
    f"{ESC}\\",
    f"{ESC}8",
    f"{ESC}7",
    f"{ESC}=",
    "plain text",
    f"{ESC}[31",
    f"{ESC}",
    f"{ESC}[31m{ESC}[0m{ESC}]11;rgb:ff/00/00\x07end",
    f"a{ESC}]0;title\x07b{ESC}Pq{ESC}\\c",
    f"{ESC}]8;;http://x{ESC}\\link{ESC}]8;;{ESC}\\",
    f"{ESC}[3",
    f"x{ESC}[31;1m{ESC}[Kbold{ESC}[0m",
    f"{ESC}]0;\x07",
    f"{ESC}P{ESC}\\",
    "line1\r\nline2\rline3\n",
    f"pre{ESC}[31m{ESC}[0m post",
    "\r\n\r\n",
    "",
    f"{ESC}[38;2;255;0;0mcolor{ESC}[0m",
    f"{ESC}]0;this {ESC} is not terminated\x07end",
]


def test_corpus_strip_parity():
    for s in ANSI_CORPUS:
        native = stream.strip_ansi(s)
        compat = stream._compat_filter_output(s)
        # strip_ansi does NOT do CRLF normalization; compare against the
        # regex-sub half of filter_output only when no \r is present.
        if "\r" not in s:
            assert native == compat, repr(s)


def test_corpus_filter_output_parity():
    for s in ANSI_CORPUS:
        native = stream.filter_output(s)
        compat = stream._compat_filter_output(s)
        assert native == compat, repr(s)


def test_filter_output_crlf():
    assert stream.filter_output("a\r\nb\rc\nd") == "a\nb\nc\nd"
    assert stream.filter_output("tail\r") == "tail\n"
    assert stream.filter_output("a\r\rb") == "a\n\nb"
    assert stream.filter_output("a\r\n\r\nb") == "a\n\nb"
    assert stream.filter_output(f"{ESC}[31mred{ESC}[0m\r\nnext") == "red\nnext"


def _oneshot_native(s):
    lp = stream.LineProcessor(strip_ansi=True, dedup_mode=0)
    out = lp.feed(s.encode("utf-8"))
    out += lp.flush()
    return "\n".join(b if isinstance(b, str) else b.decode("utf-8") for b in out)


def test_every_byte_offset_split():
    # 200-byte sample containing every escape family.
    sample = (
        f"line1{ESC}[31mred{ESC}[0m\n"
        f"{ESC}]0;title\x07osc-bel\n"
        f"{ESC}]0;title{ESC}\\osc-st\n"
        f"{ESC}Pdcs\x07dcs\n"
        f"{ESC}Pfall{ESC}[32mdcs-fall\n"
        f"{ESC}_apc\x07apc\n"
        f"{ESC}Mfe\n"
        f"{ESC}[?25lhide{ESC}[?25hshow\n"
        f"{ESC}[1 2mkept{ESC}[2J\n"
        "partial " + f"{ESC}[3" + " at eof"
    )
    while len(sample) < 200:
        sample += f"padding {ESC}[0m "
    one_shot = _oneshot_native(sample)
    assert one_shot
    data = sample.encode("utf-8")
    for split in range(len(data) + 1):
        lp = stream.LineProcessor(strip_ansi=True, dedup_mode=0)
        out = lp.feed(data[:split])
        out += lp.feed(data[split:])
        out += lp.flush()
        joined = "\n".join(out)  # shim returns str lines
        assert joined == one_shot, f"split at byte {split}"


def test_byte_by_byte_feed():
    sample = f"abc{ESC}[31mdef{ESC}]0;t\x07ghi{ESC}Pq{ESC}\\jkl{ESC}"
    data = sample.encode("utf-8")
    lp = stream.LineProcessor(strip_ansi=True, dedup_mode=0)
    out = []
    for i in range(len(data)):
        out += lp.feed(data[i : i + 1])
    out += lp.flush()
    joined = "".join(out)  # shim returns str lines
    assert joined == stream._compat_filter_output(sample)


def test_random_corpus_parity():
    rng = random.Random(42)
    esc_fragments = [
        f"{ESC}[31m", f"{ESC}[0m", f"{ESC}[?25l", f"{ESC}[?25h", f"{ESC}[2J",
        f"{ESC}]0;x\x07", f"{ESC}]0;x{ESC}\\", f"{ESC}P", f"{ESC}_", f"{ESC}M",
        f"{ESC}", f"{ESC}[", f"{ESC}[3", f"{ESC}[1 2", f"{ESC}]", f"{ESC}^",
    ]
    plain = ["text ", "line\n", "data ", "\r\n", "\r", "你 ", "😀 ", "\x07"]
    samples = []
    for _ in range(300):
        parts = []
        for _ in range(rng.randint(0, 12)):
            parts.append(rng.choice(esc_fragments) if rng.random() < 0.5 else rng.choice(plain))
        samples.append("".join(parts))
    for s in samples:
        assert stream.filter_output(s) == stream._compat_filter_output(s), repr(s[:80])
        # LineProcessor yields LINES (trailing newline dropped) — compare
        # against splitlines() of the reference filter_output.
        lp = stream.LineProcessor(strip_ansi=True, dedup_mode=0)
        out = lp.feed(s) + lp.flush()
        assert "\n".join(out) == "\n".join(stream._compat_filter_output(s).splitlines()), repr(s[:80])


def test_toggle_fallback(monkeypatch):
    monkeypatch.setenv("KIMIX_NATIVE_STREAM", "0")
    assert kimix_native.use_native("STREAM") is False
    for s in ANSI_CORPUS:
        assert stream.filter_output(s) == stream._compat_filter_output(s), repr(s)
        if "\r" not in s:
            assert stream.strip_ansi(s) == stream._compat_filter_output(s), repr(s)
    # LineProcessor falls back to _CompatLineProcessor
    lp = stream.LineProcessor(strip_ansi=True, dedup_mode=0)
    assert isinstance(lp._impl, stream._CompatLineProcessor)
    out = lp.feed(f"{ESC}[31mred{ESC}[0m\nnext") + lp.flush()
    assert out == ["red", "next"]

File: tests/unit/native/test_ansi.cpp
// Test for src/runtime/stream/ansi.h (streaming ANSI escape stripper).
// This test covers:
// - every branch of the regex: OSC (BEL/ESC\ terminated), DCS/PM/APC,
//   Fe sequences (0x40-0x5A and 0x5C-0x5F), CSI (params/intermediates/final)
// - regex backtracking fallbacks (unterminated OSC/DCS -> Fe, `\x1b[1 2m` kept)
// - partial-escape-at-eof and chunk-boundary behavior: feeding at EVERY byte
//   offset of a 200-byte sample equals one-shot processing
// - filter_output: ANSI strip + CRLF/lone-CR normalization

#include "ut/ut.hpp"
#include <runtime/stream/ansi.h>
#include <runtime/stream/line_processor.h>

#include <string>

using namespace boost::ut;
using namespace boost::ut::literals;
using namespace kimix::runtime::stream;

namespace {
std::string esc() { return std::string(1, '\x1B'); }
}

int main(int argc, char* argv[]) {
    boost::ut::detail::cfg::parse_arg_with_fallback(
        argc, const_cast<const char**>(argv));

    "one_shot_branches"_test = [] {
        // CSI SGR
        expect(eq(strip_ansi(kimix::string_view("\x1b[31mred\x1b[0m")), kimix::string("red")));
        // OSC terminated by BEL
        expect(eq(strip_ansi(kimix::string_view("\x1b]0;title\x07hello")), kimix::string("hello")));
        // OSC terminated by ESC\
        expect(eq(strip_ansi(kimix::string_view("\x1b]0;title\x1b\\hello")), kimix::string("hello")));
        // DCS terminated by BEL
        expect(eq(strip_ansi(kimix::string_view("\x1bPabc\x07x")), kimix::string("x")));
        // DCS fails at ESC+[, Fe fallback: `\x1bP` removed, abc kept
        expect(eq(strip_ansi(kimix::string_view("\x1bPabc\x1b[31mx")), kimix::string("abcx")));
        // DCS branch wins over Fe when terminated
        expect(eq(strip_ansi(kimix::string_view("\x1b_ab\x07")), kimix::string()));
        // Unterminated DCS falls back to Fe (2 bytes) at EOF
        expect(eq(strip_ansi(kimix::string_view("\x1b_")), kimix::string()));
        expect(eq(strip_ansi(kimix::string_view("\x1bP")), kimix::string()));
        expect(eq(strip_ansi(kimix::string_view("\x1b^")), kimix::string()));
        // Unterminated OSC at EOF: Fe fallback removes `\x1b]`, content kept
        expect(eq(strip_ansi(kimix::string_view("\x1b]abc")), kimix::string("abc")));
        // OSC fails at ESC+[, CSI after removed -> '\x1b]a'
        expect(eq(strip_ansi(kimix::string_view("\x1b]a\x1b[0m")), kimix::string("a")));
        // OSC fails; Fe at 2nd ESC; trailing ESC kept -> "a\x1b"
        expect(eq(strip_ansi(kimix::string_view("\x1b]a\x1b\x1b\\")), kimix::string("a\x1b")));
        // CSI ordering constraint: `\x1b[1 2m` (param space param) NOT a match
        expect(eq(strip_ansi(kimix::string_view("\x1b[1 2m")), kimix::string("\x1b[1 2m")));
        // Plain SGR, cursor hide, clear
        expect(eq(strip_ansi(kimix::string_view("\x1b[31m")), kimix::string()));
        expect(eq(strip_ansi(kimix::string_view("\x1b[?25l")), kimix::string()));
        expect(eq(strip_ansi(kimix::string_view("\x1b[2J\x1b[H")), kimix::string()));
        // Fe sequences: '(' 0x28 and '-' 0x2D are NOT in the class -> kept
        expect(eq(strip_ansi(kimix::string_view("\x1b(0")), kimix::string("\x1b(0")));
        expect(eq(strip_ansi(kimix::string_view("\x1b-")), kimix::string("\x1b-")));
        expect(eq(strip_ansi(kimix::string_view("\x1b" "8")), kimix::string("\x1b" "8")));
        expect(eq(strip_ansi(kimix::string_view("\x1b=")), kimix::string("\x1b=")));
        // Fe in class: M, \, ^(via DCS fallback), _ (via DCS fallback)
        expect(eq(strip_ansi(kimix::string_view("\x1bM")), kimix::string()));
        expect(eq(strip_ansi(kimix::string_view("\x1b\\")), kimix::string()));
        // Unterminated CSI / lone ESC at EOF -> kept
        expect(eq(strip_ansi(kimix::string_view("\x1b[31")), kimix::string("\x1b[31")));
        expect(eq(strip_ansi(kimix::string_view("\x1b")), kimix::string("\x1b")));
        // Plain text untouched
        expect(eq(strip_ansi(kimix::string_view("plain text")), kimix::string("plain text")));
        // Mixed: OSC + DCS + CSI
        expect(eq(strip_ansi(kimix::string_view("a\x1b]0;title\x07" "b\x1bPq\x1b\\c")),
                  kimix::string("abc")));
        // OSC hyperlinks
        expect(eq(strip_ansi(kimix::string_view("\x1b]8;;http://x\x1b\\link\x1b]8;;\x1b\\")),
                  kimix::string("link")));
        // Empty OSC / DCS
        expect(eq(strip_ansi(kimix::string_view("\x1b]0;\x07")), kimix::string()));
        expect(eq(strip_ansi(kimix::string_view("\x1bP\x1b\\")), kimix::string()));
        // Partial CSI at EOF followed by text
        expect(eq(strip_ansi(kimix::string_view("\x1b[3")), kimix::string("\x1b[3")));
        // SGR with ';' params and erase-line
        expect(eq(strip_ansi(kimix::string_view("x\x1b[31;1m\x1b[Kbold\x1b[0m")),
                  kimix::string("xbold")));
    };

    "chunk_boundary_split_all_offsets"_test = [] {
        // 200-byte sample containing every escape family, split at EVERY byte
        // offset: feeding {prefix}{suffix} must equal one-shot processing.
        std::string sample;
        sample += "line1\x1b[31mred\x1b[0m\n";
        sample += "\x1b]0;title\x07osc-bel\n";
        sample += "\x1b]0;title\x1b\\osc-st\n";
        sample += "\x1bPdcs\x07" "dcs\n";
        sample += "\x1bPfall\x1b[32mdcs-fall\n";
        sample += "\x1b_apc\x07" "apc\n";
        sample += "\x1bMfe\n";
        sample += "\x1b[?25lhide\x1b[?25hshow\n";
        sample += "\x1b[1 2mkept\x1b[2J\n";
        sample += "partial \x1b[3 at eof";
        // pad to >= 200 bytes
        while (sample.size() < 200u) {
            sample += "padding text \x1b[0m ";
        }

        const kimix::string one_shot = strip_ansi(kimix::string_view(sample));
        expect(!one_shot.empty());

        for (size_t split = 0; split <= sample.size(); ++split) {
            kimix::string out;
            AnsiStripper s;
            s.feed(kimix::string_view(sample.data(), split), out);
            s.feed(kimix::string_view(sample.data() + split, sample.size() - split), out);
            s.flush(out);
            expect(eq(out, one_shot)) << "split at byte " << split;
        }
    };

    "chunk_boundary_every_byte"_test = [] {
        // Feed byte-by-byte: must equal one-shot.
        std::string sample = "abc\x1b[31mdef\x1b]0;t\x07ghi\x1bPq\x1b\\jkl\x1b";
        const kimix::string one_shot = strip_ansi(kimix::string_view(sample));
        kimix::string out;
        AnsiStripper s;
        for (size_t i = 0; i < sample.size(); ++i) {
            s.feed(kimix::string_view(sample.data() + i, 1), out);
        }
        s.flush(out);
        expect(eq(out, one_shot));
        // Also: flush with no feed does not crash and emits nothing.
        kimix::string out2;
        AnsiStripper s2;
        s2.flush(out2);
        expect(out2.empty());
    };

    "filter_output_crlf"_test = [] {
        // ANSI strip + CRLF/CR normalization (tools.common.py filter_output).
        expect(eq(filter_output(kimix::string_view("a\r\nb\rc\nd")),
                  kimix::string("a\nb\nc\nd")));
        expect(eq(filter_output(kimix::string_view("\x1b[31mred\x1b[0m\r\nnext")),
                  kimix::string("red\nnext")));
        // CR at end of buffer -> LF
        expect(eq(filter_output(kimix::string_view("tail\r")), kimix::string("tail\n")));
        // Multiple CRs
        expect(eq(filter_output(kimix::string_view("a\r\rb")), kimix::string("a\n\nb")));
        expect(eq(filter_output(kimix::string_view("a\r\n\r\nb")), kimix::string("a\n\nb")));
    };

    "reset_clears_state"_test = [] {
        AnsiStripper s;
        kimix::string out;
        s.feed(kimix::string_view("\x1b[31", 4), out); // partial CSI (ESC [ 3 1)
        s.reset();
        expect(out.empty());
        s.feed(kimix::string_view("x\x1b[0m", 5), out);
        expect(eq(out, kimix::string("x")));
    };
}
